### PR TITLE
Channels

### PR DESCRIPTION
**Pull request by JanEggers**
missing: end the write loops at some point 

I didnt see a significant improvement in my benchmarks but it should perform better:

the idea is batching stuff a little:

read mutiple messages if possible before calling readasync again
write multiple messages before flushing the buffer

### CODE REVIEW

File: Bedrock.Framework/Protocols/Protocol.cs [hunk 1]
```diff
@@ -53,40 +66,25 @@ public async ValueTask<TReadMessage> ReadAsync(CancellationToken cancellationTok
...
                             var maxMessageSize = _maximumMessageSize.Value;
+                            if (buffer.Length > maxMessageSize)
+                            {
+                                inputBuffer.Complete(new InvalidDataException($"The maximum message size of {maxMessageSize}B was exceeded. The message size can be configured in AddHubOptions."));
+                            }
 
-                            if (!buffer.IsEmpty)
+                            while (reader.TryParseMessage(buffer, out consumed, out examined, out var protocolMessage))
                             {
-                                var segment = buffer;
```
Comment (davidfowl) [inline Bedrock.Framework/Protocols/Protocol.cs]: Why did you remove this logic?

File: Bedrock.Framework/Protocols/Protocol.cs [hunk 2]
```diff
@@ -53,40 +66,25 @@ public async ValueTask<TReadMessage> ReadAsync(CancellationToken cancellationTok
...
-                                return protocolMessage;
+                                await inputBuffer.WriteAsync(protocolMessage);
+                                buffer = buffer.Slice(consumed);
                             }
                         }
                         else
                         {
                             // We give the parser a sliding window of the default message size
                             var maxMessageSize = _maximumMessageSize.Value;
+                            if (buffer.Length > maxMessageSize)
```
Comment (davidfowl) [inline Bedrock.Framework/Protocols/Protocol.cs]: This logic is incorrect.

File: Bedrock.Framework/Protocols/Protocol.cs (hunk 1 above)
Comment (JanEggers) [reply Bedrock.Framework/Protocols/Protocol.cs]: the same check is in line 62 so i think its redundant

Comment (JanEggers) [reply Bedrock.Framework/Protocols/Protocol.cs]: regarding the other comment: I reverted the change in the branch with msg size limit seems I overlooked something while trying to simplify things
